macrosynergy/pnl/historic_portfolio_volatility.py

```diff
@@ -210,69 +329,62 @@ def _calculate_portfolio_volatility(
     # TODO convert frequencies
     list_vcv: List[pd.DataFrame] = []
     list_pvol: List[Tuple[pd.Timestamp, np.float64]] = []
+    first_starts = _get_first_usable_date(
+        pivot_returns=pivot_returns,
+        pivot_signals=pivot_signals,
+        rebal_dates=rebal_dates,
+        est_freqs=est_freqs,
+        lback_periods=lback_periods,
+        nan_tolerance=nan_tolerance,
+    )
+
     for td in rebal_dates:
-        window_df = pivot_returns.loc[pivot_returns.index <= td]
-        dict_vcv: Dict[str, pd.DataFrame] = {
-            freq: estimate_variance_covariance(
-                piv_ret=_downsample_returns(window_df, freq=freq).iloc[
-                    -get_max_lookback(lb=lb, nt=nan_tolerance) :
-                ],
-                lback_periods=lb,
-                remove_zeros=remove_zeros,
-                weights_func=weights_func,
-                half_life=hl,
+        avails = first_starts[first_starts <= td].index.tolist()
```

Comment:
We need to check there is no signals associated with not available cross sections in prior sets of the sample.